File: java/roman-numerals/src/com/numerals/roman/Conversion.java
/**
 * @brief Roman numerals code kata.
 */
package com.numerals.roman;

import java.util.Map;
import java.util.TreeMap;

/**
 * @brief Roman numerals conversion functions.
 */
public class Conversion
{
  private static final TreeMap<Integer, String> alphabet = createAlphabet();

  /**
   * @brief Creates the Roman alphabet.
   *
   * @return The Roman alphabet value-symbol map.
   */
  private static TreeMap<Integer, String> createAlphabet()
  {
    TreeMap<Integer, String> result = new TreeMap<Integer, String>();

    result.put(1000, "M"); result.put(900, "CM"); result.put(500, "D"); result.put(400, "CD");
    result.put( 100, "C"); result.put( 90, "XC"); result.put( 50, "L"); result.put( 40, "XL");
    result.put(  10, "X"); result.put(  9, "IX"); result.put(  5, "V"); result.put(  4, "IV");
    result.put(   1, "I");

    return result;
  }

  /**
   * @brief Converts a Arabic number into it's Roman number representation.
   *
   * @param number The number to convert.
   *
   * @return The Roman number as a string.
   */
  public static String toRoman(int number)
  {
    String result = "";

    while (number > 0)
    {
      Map.Entry<Integer, String> digit = alphabet.floorEntry(number);

      number -= digit.getKey();
      result += digit.getValue();
    }

    return result;
  }
}

File: csharp/roman-numerals/Roman.Numerals/Conversion.cs
// <copyright file="Conversion.cs" company="Alejandro Claro">
//     Copyright (c) Alejandro Claro. All rights reserved.
// </copyright>
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roman.Numerals
{
  /// <summary>
  /// Roman numerals conversion functions.
  /// </summary>
  public class Conversion
  {
    private static readonly IComparer<uint> reverseComparer = Comparer<uint>.Create((x, y) => y.CompareTo(x));
    private static readonly SortedDictionary<uint, string> alphabet = new SortedDictionary<uint, string>(reverseComparer)
    {
      { 1000,  "M" }, { 900, "CM" }, { 500,  "D" }, { 400, "CD" },
      {  100,  "C" }, {  90, "XC" }, {  50,  "L" }, {  40, "XL" },
      {   10,  "X" }, {   9, "IX" }, {   5,  "V" }, {   4, "IV" },
      {    1,  "I" }, {   0,   "" }
    };

    /// <summary>
    /// Converts an arabic number into it's Roman number representation.
    /// </summary>
    ///
    /// <param name="number">The arabic number.</param>
    ///
    /// <returns>The Roman number as a string.</returns>
    public static string ToRoman(uint number)
    {
      string result = "";

      while (number > 0)
      {
        var digit = alphabet.First(x => (number >= x.Key));

        number -= digit.Key;
        result += digit.Value;
      }

      return result;
    }
  }
}

